File: platform-runtime/service-shells/engine-core-service/lifecycle.ts
/**
 * Engine-core service shell lifecycle.
 * Independent startup/shutdown for this process boundary.
 */

export type LifecycleHook = () => Promise<void>;

const startupHooks: LifecycleHook[] = [];
const shutdownHooks: LifecycleHook[] = [];
let shuttingDown = false;

export function registerStartupHook(hook: LifecycleHook): void {
  startupHooks.push(hook);
}

export function registerShutdownHook(hook: LifecycleHook): void {
  shutdownHooks.push(hook);
}

export async function runStartup(): Promise<void> {
  for (const hook of startupHooks) {
    await hook();
  }
}

export async function runShutdown(): Promise<void> {
  if (shuttingDown) return;
  shuttingDown = true;
  for (const hook of [...shutdownHooks].reverse()) {
    try {
      await hook();
    } catch (e) {
      console.error('Shutdown hook error:', e);
    }
  }
}

export function isShuttingDown(): boolean {
  return shuttingDown;
}

File: services/engine-trust-service/app/lifecycle.py
"""Startup/shutdown lifecycle for engine-trust-service."""
from typing import Callable, List

_startup_hooks: List[Callable[[], None]] = []
_shutdown_hooks: List[Callable[[], None]] = []


def on_startup(fn: Callable[[], None]) -> None:
    _startup_hooks.append(fn)


def on_shutdown(fn: Callable[[], None]) -> None:
    _shutdown_hooks.append(fn)


async def run_startup() -> None:
    for fn in _startup_hooks:
        fn()


async def run_shutdown() -> None:
    for fn in reversed(_shutdown_hooks):
        fn()
